### ISSUE TITLE
Some additional concerns and confusion

### ISSUE THREAD

**Issue by JaceJu-frog** [Ruifeng-Tan/BatteryLife#7]
> Thank you for your interest in our work. Sorry for the inconvenience caused. Please find the point-by-point responses as follows:
> - Q1: Are these `cal_for_test.json` files unnecessary for execution?
> 
> A1: Sorry for missing these files. We have uploaded these files to the dataset/seen_unseen_labels folder of the repo. I clarify its role in the answer for Q2.
> 
> - Q2: I don't understand the role of `unseen_id`. Test/train/val splits typically don't require unseen IDs (models naturally shouldn't know the test set). `BatteryML` also doesn’t use this variable, and the dataset splits are explicitly defined in `data_provider/data_split_recorder.py`. Could you clarify its purpose?
> 
> A2: You are right. If you simply want to train the model, you do not need the `unseen_id`. It is viable to randomly set `total_seen_unseen_IDs` for the training set (line 362 in `data_loader.py`). You can also see in the `run_main.py` that the `seen_unseen_ids` is just loaded but affects nothing during the model training loop. The `unseen_id` and `cal_for_test.json` files are included here to summarize the model performance on aging conditions that are seen and unseen in the training data (Line 376 in `run_main.py`, Table 5 of [BatteryLife paper](https://arxiv.org/pdf/2502.18807)). The results will be printed when the training is finished. If you don't care about the model performance on the seen and unseen aging conditions, you can also set the `total_seen_unseen_IDs` like `line 362 in data_loader.py` for the testing set. As for  `dataset splits`, they were randomly generated and recorded in `data_provider/data_split_recorder.py`. This is to ensure consistent dataset splits for future studies that want to compare their proposed methods with the benchmark methods of this work.
> 
> - Q3: The presence of seen_unseen_label in models/dummy.py suggests legacy code. Are these unrelated to normal training/evaluation workflows?
> 
> A3: Sorry for the confusion caused. We have included the note for this model in https://github.com/Ruifeng-Tan/BatteryLife/blob/main/assets/Model_training.md now.
> 
> If you have further comments regarding this issue, please feel free to post them here. 

Hi Tan ,

First, thank you for your great work and the contribution it has made to the benchmark community. I’ve been following the project for a few months, and it’s inspiring to see the star count rise past 110—already a respectable number in the battery life-prediction field. Your work has given me many ideas for my own research.

---

### 1. Questions about the *iTransformer* training setup

```text
before enc_embedding           → (16, 30000, 3)
after  enc_embedding           → (16, 3, 128)
```

The implementation follows the original paper’s of itransformer's idea of treating each **variable** (here: voltage, current, capacity) as a single token.
However, jumping directly from **30 000** to **128** dimensions feels abrupt—especially when the original GitHub repo mentions a maximum input length of 720. That’s a >40× compression in one linear layer. Wouldn’t a treatment similar to **CPTransformer** be more appropriate?

Example (pseudo-shapes):

```
Input            (B, 100, 3, 300)
↓ reshape        (B, 20, 3, 1500)
↓ linear         (B, 20, 3, 128)
↓ MLP × N
↓ Encoder
→ output         (B, 1)  # life prediction
```

Splitting the five cycles’ information( Five serves as  example for this hyperparameter, in fact 1,2,10,20 is also acceptable ) into separate channels (3, 300*5) seems more reasonable.

> what's more ,the last fc layer (12800,1) is also worthy of further discussion, such a huge layer.

---

### 2. Confusion about the `curve_attn_mask` in **CPTransformer**

While training (e.g. `CPtransformer.py`, line 65), there is a comment `# set the unseen data as zeros`.
Running `CPtransformer.sh` with `MIX_large` and default settings, I noticed that `forward` receives a mask of shape **(batch, 100)**, and not all entries are 1. Could you explain the purpose of this mask? My experience is that masks are only needed for variable-length sequences. Why hide part of the first 100 cycles from the model?

---

### 3. Suggestions

1. **Prediction target** – Time-series models usually forecast an entire curve, not a single point. Have you considered adding the task “predict the last cycle (defined at 80 % of maximum capacity)”?
2. **New metric** – Besides MAPE, accuracy could be a useful evaluation metric. In code I only see MSE used for back-prop; perhaps increasing the weight of MAPE/accuracy losses would be helpful.
3. **Definition of `unseen_id` / aging condition** – Am I correct that *aging condition* is effectively the charge/discharge protocol? i.e., whenever the protocol differs, it is deemed a different aging condition?
4. **Additional dataset** – I found a potentially valuable set:

   *Dataset:* [https://figshare.com/articles/dataset/eVTOL\_Battery\_Dataset/14226830/1?file=26855069](https://figshare.com/articles/dataset/eVTOL_Battery_Dataset/14226830/1?file=26855069)
   *Paper:* [https://www.nature.com/articles/s41597-023-02180-5](https://www.nature.com/articles/s41597-023-02180-5)

---

### 4. Possible PR

I’m happy to open a PR that formalizes the above, but it may take some time: my current work involves proprietary battery data and I can’t devote full attention immediately.

---

Thanks again for the impressive benchmark and your continued effort!

**Comment by JaceJu-frog** [comment]
I have learned the role of attention mask in [issue6](https://github.com/Ruifeng-Tan/BatteryLife/issues/6)

**Comment by Ruifeng-Tan** [comment]
Thank you for your questions and suggestions. Please find point-by-point responses below:

**Comment by Ruifeng-Tan** [comment]
Thank you for your questions and suggestions. Please find point-by-point responses below:

**1. Questions about the iTransformer training setup**

In this work, as you have mentioned, we faithfully follow the original implementation of iTransformer. As you have suggested, it might be more suitable to split the information of $$N$$ cycles. From my view, this becomes another kind of patching instead of the variate tokens. During the CyclePatch development, we have discussions within the team to implement extended CyclePatch, i.e., patching $$N$$ cycles with $$N\geq 1$$. However, we did not obtain clear improvements in some very preliminary tests, and the KDD deadline was around the corner. Therefore, we decided to simplify the story and use the CyclePatch with only one cycle patched. You can still try in this direction.

As for the output layer, we follow the implementations from [Time-Series-Library](https://github.com/thuml/Time-Series-Library). It is possible that some improvements can be conducted.

**2. Confusion about the curve_attn_mask in CPTransformer**

Sorry for the confusion. I am glad that you have learned its role.

**3. Suggestions**
Thank you for your suggestions.

- Predicted target. Forecasting the battery SOH degradation curve is an important task in the AI4Battery community. If you are interested in this task, please feel free to launch a PR. During the implementation, I kindly suggest you take care of the following points:
 
   - Forecasting the SOH degradation curve is kind of different from the common time series forecasting settings in the ML literature (e.g., iTransformer and PatchTST papers). In the iTransformer settings, researchers forecast the future states using the historical states (e.g., forecast future temperature using historical temperature). However, we are interested in using historical voltage, current, and capacity variables to forecast the future SOH degradation trajectory. In addition, both the inputs and outputs can be times longer than the longest settings in the ML literature.
   - The battery SOH degradation trajectories might contain random SOH changes due to the rest performance tests (RPTs). During the RPTs, researchers may conduct some characterizations or cycling with special protocols. Although we have removed the RPT cycles, the SOHs of the cycles following the RPT cycles are still affected by the RPTs. This makes the SOH of some regions almost unpredictable.

- New metric. It would be interesting to optimize the model with a loss tailored for MAPE, but I do not understand how to optimize accuracy (a classification metric) for a regression task.
- Definition of unseen_id / aging condition. Aging condition is affected by many aging factors in addition to the charge/discharge protocol. In this work, we considered 9 aging factors, and you can find the aging factors in Section 2.1 of [our paper](https://arxiv.org/abs/2502.18807).
- Additional dataset. This is an interesting dataset. We would be happy to review the PR if you are interested in standardizing it.

Thanks again for your questions and suggestions.

**Comment by JaceJu-frog** [comment]
- Ok you are right ,as the 15-accuracy may not be directly used as  loss in gradient descent, while a softed version may count. I am thinking about changing loss in curve prediction (my work), where the curve’s numerical values and its overall trend are important, so we can’t rely on MSE loss alone for training.
- If I will predict curve, I may remove rest performance tests too, keep the same standard in selecting cycles .
- Hope I may finish my work and give you a PR for points mentioned .

> By the way, was the CALB dataset generated and tested in your lab? Its 60 Ah capacity is considerably larger than that of any other dataset referenced in this repository.
